src/wodan/wodan_lru.ml

```diff
@@ -0,0 +1,46 @@
+module Make (K : Hashtbl.HashedType) (V : Lru.Weighted) = struct
+  module LRU = Lru.M.Make (K) (V)
+
+  type t = LRU.t
+
+  type key = K.t
+
+  type value = V.t
+
+  exception Already_cached of key
+
+  let get_opt t k = LRU.find ~promote:true k t
+
+  let get t k =
+    match get_opt t k with
+    | None ->
+        raise Not_found
+    | Some x ->
+        x
+
+  let peek_opt t k = LRU.find ~promote:false k t
+
+  let peek t k =
+    match peek_opt t k with
+    | None ->
+        raise Not_found
+    | Some x ->
+        x
+
+  let xadd t k v =
+    if LRU.mem k t then raise @@ Already_cached k else LRU.add k v t
```

Comment:
`@@` is really necessary?

Reply:
You mean `raise (Already_cached k)` would be better ?

Reply:
I tend to avoid `@@` as well if possible. This gives more work to the optimiser (see https://discuss.ocaml.org/t/benchmarking-bucklescript-vs-js-of-ocaml/3191/15) and I found it generally less readable too.

Reply:
I did not know that, it's a good point.